File: SKOS-Android/app/src/main/java/com/smaily/skos/engine/EngineRegistry.kt
package com.smaily.skos.engine

import com.smaily.skos.common.ComponentType
import java.util.concurrent.ConcurrentHashMap

/**
 * ---------------------------------------------------------
 * SKOS Professional
 * Engine Registry
 * ---------------------------------------------------------
 * Central registry of all SKOS engines.
 * ---------------------------------------------------------
 */
class EngineRegistry {

    /**
     * Registered engines.
     */
    private val engines =
        ConcurrentHashMap<ComponentType, Engine>()

    /**
     * Register engine.
     */
    fun register(engine: Engine): Boolean {

        return engines.putIfAbsent(
            engine.type,
            engine
        ) == null
    }

    /**
     * Remove engine.
     */
    fun unregister(type: ComponentType): Engine? {

        return engines.remove(type)
    }

    /**
     * Get engine.
     */
    fun get(type: ComponentType): Engine? {

        return engines[type]
    }

    /**
     * Check existence.
     */
    fun contains(type: ComponentType): Boolean {

        return engines.containsKey(type)
    }

    /**
     * All engines.
     */
    fun getAll(): List<Engine> {

        return engines.values.toList()
    }

    /**
     * Number of engines.
     */
    fun size(): Int {

        return engines.size
    }

    /**
     * Registry empty?
     */
    fun isEmpty(): Boolean {

        return engines.isEmpty()
    }

    /**
     * Remove everything.
     */
    fun clear() {

        engines.clear()
    }
}

File: mission-control/assets/js/integration/engine-registry.js
/**
 * ============================================================
 * SKOS - Smaily Knowledge Operating System
 * Engine Registry
 * ------------------------------------------------------------
 * File      : engine-registry.js
 * Operation : OP-021
 * Build     : BUILD-000421
 * Version   : 1.0.0
 * Status    : ACTIVE
 * ============================================================
 *
 * Mission:
 * Central registry for all SKOS engines.
 *
 * Responsibilities:
 * - Register engines
 * - Unregister engines
 * - Discover engines
 * - Search engines
 * - Categorize engines
 * - Maintain engine metadata
 * - Validate registry integrity
 *
 * ============================================================
 */

class EngineRegistry {

    constructor(config = {}) {

        this.name = "EngineRegistry";
        this.version = "1.0.0";

        this.config = config;

        this.initialized = false;
        this.running = false;

        this.engines = new Map();

        this.categories = new Map();

        this.statistics = {

            enginesRegistered: 0,
            enginesRemoved: 0,
            engineLookups: 0,
            categoryAssignments: 0,
            validations: 0

        };

    }

    /**
     * Initialize
     */
    initialize() {

        if (this.initialized) {

            return true;

        }

        this.initialized = true;

        return true;

    }

    /**
     * Execute
     */
    execute() {

        if (!this.initialized) {

            this.initialize();

        }

        this.running = true;

        return true;

    }

    /**
     * Shutdown
     */
    shutdown() {

        this.running = false;

        return true;

    }

    /**
     * Register Engine
     */
    register(engine = {}) {

        if (!engine.name) {

            return false;

        }

        this.engines.set(

            engine.name,

            {

                ...engine,

                registeredAt: new Date()

            }

        );

        this.statistics.enginesRegistered++;

        return true;

    }

    /**
     * Remove Engine
     */
    unregister(name) {

        if (!this.engines.has(name)) {

            return false;

        }

        this.engines.delete(name);

        this.statistics.enginesRemoved++;

        return true;

    }

    /**
     * Find Engine
     */
    find(name) {

        this.statistics.engineLookups++;

        return this.engines.get(name) || null;

    }

    /**
     * Register Category
     */
    assignCategory(name, category) {

        if (!this.engines.has(name)) {

            return false;

        }

        this.categories.set(name, category);

        this.statistics.categoryAssignments++;

        return true;

    }

    /**
     * Get Engines By Category
     */
    getByCategory(category) {

        const result = [];

        this.categories.forEach(

            (value, key) => {

                if (value === category) {

                    result.push(

                        this.engines.get(key)

                    );

                }

            }

        );

        return result;

    }

    /**
     * Get All Engines
     */
    getAll() {

        return Array.from(

            this.engines.values()

        );

    }

    /**
     * Validate Registry
     */
    validate() {

        this.statistics.validations++;

        return {

            valid: true,

            totalEngines:

                this.engines.size

        };

    }

    /**
     * Health Check
     */
    healthCheck() {

        return {

            registry: this.name,

            version: this.version,

            initialized: this.initialized,

            running: this.running,

            engines:

                this.engines.size,

            categories:

                this.categories.size,

            statistics:

                this.statistics

        };

    }

    /**
     * Reset
     */
    reset() {

        this.engines.clear();

        this.categories.clear();

        this.initialized = false;

        this.running = false;

        this.statistics = {

            enginesRegistered: 0,
            enginesRemoved: 0,
            engineLookups: 0,
            categoryAssignments: 0,
            validations: 0

        };

    }

}

/**
 * Export
 */

if (typeof module !== "undefined") {

    module.exports = EngineRegistry;

}

if (typeof window !== "undefined") {

    window.EngineRegistry = EngineRegistry;

}
